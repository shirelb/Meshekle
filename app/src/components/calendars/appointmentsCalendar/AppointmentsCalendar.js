import React, {Component} from 'react';
import {ScrollView, StyleSheet} from 'react-native';
import {Calendar, LocaleConfig} from 'react-native-calendars';
import '../localConfig';
import moment from 'moment';
import phoneStorage from "react-native-simple-store";
import appointmentsStorage from "../../../storage/appointmentsStorage";
import {APP_SOCKET} from "../../../shared/constants";
import AppointmentsDayInfo from "../../appointments/AppointmentsDayInfo";
import serviceProvidersStorage from "../../../storage/serviceProvidersStorage";

export default class AppointmentsCalendar extends Component {
    constructor(props) {
        super(props);

        this.state = {
            markedDates: {},
            selectedDate: '',
            dateModalVisible: false,
            expanded: {}
        };

        this.userHeaders = {};
        this.userId = null;
    }

    componentDidMount() {
        phoneStorage.get('userData')
            .then(userData => {
                this.userHeaders = {
                    'Authorization': 'Bearer ' + userData.token
                };
                this.userId = userData.userId;
                this.loadAppointments();
            });

        APP_SOCKET.on("getUserAppointments", this.loadAppointments.bind(this));
    }

    componentWillUnmount() {
        APP_SOCKET.off("getUserAppointmentsâ€š");
    }

    loadAppointments = () => {
        appointmentsStorage.getUserAppointments(this.userId, this.userHeaders)
            .then(response => {
                let markedDates = {};

                response.data.forEach(appointment => {
                    const date = moment(appointment.startDateAndTime).format('YYYY-MM-DD');
                    if (markedDates[date] === undefined || markedDates[date] === null) {
                        markedDates[date] = {marked: true, selected: false, appointments: []};
                    }
                    markedDates[date].appointments.push(appointment);
                });

                this.setState({
                    markedDates: markedDates
                });
            })
    }

    afterCloseModalShowSelectDay = () => {
        let updatedMarkedDates = this.state.markedDates;

        const selectedDay = updatedMarkedDates[this.state.selectedDate];
        selectedDay.selected = true;
        selectedDay.color = 'blue';
        updatedMarkedDates[selectedDay] = selectedDay;

        this.setState({markedDates: updatedMarkedDates, dateModalVisible: false})
    };

    onDaySelect = (day) => {
        let updatedMarkedDates = this.state.markedDates;

        if (this.state.selectedDate !== '') {
            let lastMarkedDate = updatedMarkedDates[this.state.selectedDate];
            lastMarkedDate.selected = false;
            updatedMarkedDates[this.state.selectedDate] = lastMarkedDate;
        }

        const selectedDay = moment(day.dateString).format('YYYY-MM-DD');
        if (updatedMarkedDates[selectedDay] === undefined)
            updatedMarkedDates[selectedDay] = {marked: true, selected: false, appointments: []};
        let newMarkedDate = updatedMarkedDates[selectedDay];
        newMarkedDate.selected = true;
        newMarkedDate.color = 'blue';
        updatedMarkedDates[selectedDay] = newMarkedDate;

        let expanded = {};
        if (updatedMarkedDates[selectedDay].appointments.length > 0)
            updatedMarkedDates[selectedDay].appointments.forEach(appointment => {
                serviceProvidersStorage.getServiceProviderUserDetails(appointment.AppointmentDetail.serviceProviderId, this.userHeaders)
                    .then(user => {
                        appointment.serviceProviderFullname = user.data.fullname;

                        expanded[appointment.appointmentId] = false;

                        this.setState({
                            selectedDate: moment(day.dateString).format('YYYY-MM-DD'),
                            markedDates: updatedMarkedDates,
                            dateModalVisible: true,
                            expanded: expanded,
                        });
                    });
            });
        else
            this.setState({
                selectedDate: moment(day.dateString).format('YYYY-MM-DD'),
                markedDates: updatedMarkedDates,
                dateModalVisible: true,
                expanded: expanded,
            });
    };


    render() {
        LocaleConfig.defaultLocale = 'il';

        return (
            <ScrollView>
                <Calendar
                    markedDates={this.state.markedDates}
                    onDayPress={this.onDaySelect}
                    onDayLongPress={this.onDaySelect}
                    style={styles.calendar}
                    // Month format in calendar title. Formatting values: http://arshaw.com/xdate/#Formatting
                    // monthFormat={'yyyy MM'}
                    // Handler which gets executed when visible month changes in calendar. Default = undefined
                    // onMonthChange={(month) => {
                    //     console.log('month changed', month)
                    // }}
                    // Replace default arrows with custom ones (direction can be 'left' or 'right')
                    // renderArrow={(direction) => (<Arrow/>)}
                    theme={{
                        backgroundColor: '#ffffff',
                        calendarBackground: '#ffffff',
                        textSectionTitleColor: '#b6c1cd',
                        selectedDayBackgroundColor: '#00adf5',
                        selectedDayTextColor: '#ffffff',
                        todayTextColor: '#00adf5',
                        dayTextColor: '#2d4150',
                        textDisabledColor: '#d9e1e8',
                        dotColor: '#00adf5',
                        selectedDotColor: '#ffffff',
                        arrowColor: 'orange',
                        monthTextColor: 'blue',
                        textDayFontFamily: 'monospace',
                        textMonthFontFamily: 'monospace',
                        textDayHeaderFontFamily: 'monospace',
                        textMonthFontWeight: 'bold',
                        textDayFontSize: 16,
                        textMonthFontSize: 16,
                        textDayHeaderFontSize: 16
                    }}
                />

                <AppointmentsDayInfo
                    dateModalVisible={this.state.dateModalVisible}
                    selectedDate={this.state.selectedDate}
                    onAppointmentRequestPress={this.props.onAppointmentRequestPress}
                    loadAppointments={this.loadAppointments}
                    markedDates={this.state.markedDates}
                    expanded={this.state.expanded}
                    afterCloseModalShowSelectDay={this.afterCloseModalShowSelectDay}
                    userHeaders={this.userHeaders}
                />
            </ScrollView>
        );
    }
}

const styles = StyleSheet.create({
    calendar: {
        borderTopWidth: 2,
        // marginTop: 5,
        paddingTop: 5,
        borderBottomWidth: 2,
        borderColor: '#eee',
        height: 350,
    },
    text: {
        textAlign: 'center',
        borderColor: '#bbb',
        padding: 10,
        backgroundColor: '#eee'
    },
});