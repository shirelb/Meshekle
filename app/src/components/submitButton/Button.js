import React, {Component} from 'react';
import {StyleSheet, Text, TouchableOpacity} from "react-native";
import PropTypes from 'prop-types'
import colors from "../../shared/colors";


export default class Button extends Component {
    render() {
        const {label, onPress} = this.props;
        return (
            <TouchableOpacity style={Object.assign({},
                styles.container,
                {backgroundColor: this.props.color ? this.props.color : colors.DODGER_BLUE})}
                              onPress={onPress}>
                <Text style={styles.text}>{label}</Text>
            </TouchableOpacity>
        );
    }
}

const styles = StyleSheet.create({
    container: {
        width: "100%",
        alignItems: "center",
        justifyContent: "center",
        marginBottom: 12,
        paddingVertical: 12,
        borderRadius: 4,
        borderWidth: StyleSheet.hairlineWidth,
        borderColor: "rgba(255,255,255,0.7)"
    },
    // containerBGcolor: {
    //     backgroundColor: colors.DODGER_BLUE,
    // },
    text: {
        color: colors.WHITE,
        textAlign: "center",
        height: 20
    }
});

Button.propTypes = {
    label: PropTypes.string.isRequired,
    onPress: PropTypes.func.isRequired,
};