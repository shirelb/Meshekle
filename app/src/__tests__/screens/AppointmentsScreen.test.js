import React, {Component} from 'react';
import {StyleSheet, View} from 'react-native';
import AppointmentsCalendar from "../../components/calendars/appointmentsCalendar/AppointmentsCalendar";
import Button from "../../components/submitButton/Button"

export default class AppointmentsScreen extends Component {

    onAppointmentRequestPress = (selectedDate = '') => {
        this.props.navigation.navigate('AppointmentRequest', {selectedDate: selectedDate})
    };

    onMyAppointmentRequestsPress = () => {
        this.props.navigation.navigate('UserAppointmentRequests',{ onAppointmentRequestPress:this.onAppointmentRequestPress})
    };


    render() {
        return (
            <View>
                <Button
                    label="בקשות התורים שלי"
                    onPress={this.onMyAppointmentRequestsPress.bind(this)}
                />

                <Button
                    label="בקש תור חדש"
                    onPress={this.onAppointmentRequestPress.bind(this)}
                />

                <AppointmentsCalendar
                    onAppointmentRequestPress={this.onAppointmentRequestPress}
                />
            </View>
        );
    }
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        // justifyContent: 'center',
        // alignItems: 'center'
    },
    buttonStyle:{
        marginTop: 50,
    }
});

