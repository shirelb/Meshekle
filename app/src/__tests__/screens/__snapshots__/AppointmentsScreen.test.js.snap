// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppointmentsScreen should match snapshot 1`] = `
<View>
  <Button
    label="בקשות התורים שלי"
    onPress={[Function]}
  />
  <Button
    label="בקש תור חדש"
    onPress={[Function]}
  />
  <AppointmentsCalendar
    onAppointmentRequestPress={[Function]}
  />
</View>
`;
